app/helpers/quadicon_helper.rb

```diff
@@ -69,95 +69,145 @@ def img_for_host_vendor(item)
-              )
-            else
-              link_to(truncate_for_quad(name), nil, h(name))
-            end
+    key = case quadicon_model_name(item)
+          # when "MiqCimInstance"       then "evm_display_name"
+          when "ConfiguredSystem"     then "hostname"
+          when "ServiceResource"      then "resource_name"
+          when "ConfigurationProfile" then "description"
+          when "EmsCluster"           then "v_qualified_desc"
           else
-            link_to(
-              truncate_for_quad(name),
-              {:action => 'x_show', :id => controller.send(:list_row_id, row)},
-              "data-miq_sparkle_on"  => true,
-              "data-miq_sparkle_off" => true,
-              :title                 => name,
-              "data-method"          => :post,
-              :remote                => true
-            )
-          end
-        else
-          if row['evm_display_name']
-            link_to(truncate_for_quad(row['evm_display_name']), url_for_db(db, "show"), :title => h(row['evm_display_name']))
-          elsif row['key']
-            link_to(truncate_for_quad(row['key']), url_for_db(db), :title => h(row['key']))
-          else
-            link_to(truncate_for_quad(row['name']), url_for_db(db, "show", item), :title => h(row['name']))
+            %w(evm_display_name key name).detect { |k| row[k] }
           end
+
+    truncate_for_quad(row[key])
+  end
+
+  def quadicon_build_label_url(item, row)
+    if quadicon_in_explorer_view?
+      quadicon_build_explorer_url(item, row)
+    else
+      url_for_db(quadicon_model_name(item))
+    end
+  end
+
+  def quadicon_build_explorer_url(item, row)
+    attrs = default_url_options
+
+    if quadicon_try_vm_quad_attributes?
+      if quadicon_vm_attributes(item)
+        attrs.merge!(quadicon_vm_attributes(item))
+      end
+    else
+      attrs[:action]  = 'x_show'
+      attrs[:id]      = ApplicationRecord.compress_id(row['id'])
+    end
+
+    url_for(attrs)
+  end
+
+  def quadicon_build_label_options(item, row)
+    options = {:title => quadicon_label_content(item, row)}
+
+    if quadicon_render_for_policy_sim?
+      options[:title] = _("Show policy details for %s") % row['name']
+    end
+
+    if quadicon_in_explorer_view?
+      if quadicon_try_vm_quad_attributes?
+        if quadicon_vm_attributes_present?(item)
+          options["data-miq_sparkle_on"] = true
+          options["data-miq_sparkle_off"] = true
         end
+      else
+        options["data-miq_sparkle_on"] = true
+        options["data-miq_sparkle_off"] = true
+        options[:data] = {:method => :post, :remote => true}
       end
     end
+
+    options
+  end
+
+  def quadicon_vm_attributes(item)
+    @quad_vm_attrs ||= vm_quad_link_attributes(item)
+  end
+
+  def quadicon_vm_attributes_present?(item)
+    quadicon_vm_attributes(item) && !quadicon_vm_attributes(item).empty?
+  end
+
+  def quadicon_in_embedded_view?
+    !!@embedded
```

Comment:
I'd love to prevent mixing new instance variables, such as `@quad_vm_attrs` (probably belonging to some class that handles the quadicon rendering) with the controller instance variables, that we already have here (such as @embedded and @policy_sim).

I'd prefer to introduce some `mode` or expand the `options[:typ]` to be that `mode` and have that express the context in which the quadicon in rendered. It would be explicitly passed when instantiating some `QuadiconRenderer` or something instead of sharing controller instance variables.

In context of:
https://github.com/ManageIQ/manageiq/pull/7623/files#diff-455dafb9cdb7dca9429d664b2ee343c7R18

We will need to render quadicons for the JSON that is fed into the GTL component. We would need to handle the `mode` there too.